README


This archive contains the distribution MooseX-StrictConstructor,
version 0.18:

  Make your object constructors blow up on unknown attributes

This software is Copyright (c) 2012 by Dave Rolsky.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)




Changes
0.18     2012-01-10

- A small internal tweak which prevents the MooseX::ClassCompositor tests from
  failing.


0.17     2012-01-04

- Added a small optimization for an immutable constructor when used with Moose
  2.00+. The list of valid attributes is only generated once, rather than each
  time the constructor is run. Reported by Chip Salzenberg. RT #73664.


0.16     2011-04-22

- Don't wrap BUILDALL, since this breaks if the object doesn't inherit from
  Moose::Object (doy).


0.15     2011-04-13

- The fix in 0.14 only worked with Moose 2.0, causing failures with earlier
  versions.


0.14     2011-04-13

- Using extends caused strictness to be lost. Fixed by Karen Etheridge.


0.13     2011-02-22

- This release provides forward compatibility with Moose 1.99+. It will still
  work with Moose 1.23 as well. Most of the work was done by Jesse Luehrs.


0.12     2010-10-29

- Switched tests to use Test::Fatal instead of Test::Exception.


0.11     2010-09-09

- Added a missing test dep on Test::Exception. Reported by Phillip Moore. RT
  #61205.


0.10     2010-07-16

- Ignore __INSTANCE__ when passed to the constructor. Patch by Jesse
  Luehrs. RT #59236.

- All init args are quoted in the inlined constructor.


0.09     2010-07-14

- Use modern Moose APIs, to avoid warnings with the next Moose release.


0.08     2009-04-07

- Make this module work with Moose 0.73_01+.


0.07     2008-09-01

- No code changes, just a stable release for Moose 0.56.


0.06_02  2008-08-30

* This module no longer exports Moose's sugar. Just use Moose instead.

- Various doc fixes to reflect changes since 0.06.


0.06_01  2008-08-30

- Rewrote this distro's functionality as roles to take advantage of
  Moose::Util::MetaRole. Less code, plays better with other
  extensions.


0.06     2008-06-27

- Changes to the immutable bits to work with the latest Moose (0.51).


0.05     2008-04-18

- The fix in 0.04 only worked for non-immutable classes.


0.04     2008-04-18

- This module did not respect the init_arg attribute setting, and used
  the attribute name instead. Reported by Matt Trout. RT #34507.


0.03     2007-11-20

- Require Moose 0.26, since that's first version to offer init_meta(),
  apparently.


0.02     2007-11-15

- Moose was missing from the prereq list. Reported by Slaven Rezic.

- Version 0.01 did not work after a class was made immutable.


0.01     2007-11-14

- First version, released on an unsuspecting world.


lib/MooseX/StrictConstructor.pm
package MooseX::StrictConstructor;
{
  $MooseX::StrictConstructor::VERSION = '0.18';
}

use strict;
use warnings;

use Moose 0.94 ();
use Moose::Exporter;
use Moose::Util::MetaRole;

use MooseX::StrictConstructor::Trait::Class;
use MooseX::StrictConstructor::Trait::Method::Constructor;;

my %metaroles = (
    class => ['MooseX::StrictConstructor::Trait::Class'],
);

$metaroles{constructor}
    = ['MooseX::StrictConstructor::Trait::Method::Constructor']
    if $Moose::VERSION >= 1.9900;

Moose::Exporter->setup_import_methods( class_metaroles => \%metaroles );

1;

# ABSTRACT: Make your object constructors blow up on unknown attributes



=pod

=head1 NAME

MooseX::StrictConstructor - Make your object constructors blow up on unknown attributes

=head1 VERSION

version 0.18

=head1 SYNOPSIS

    package My::Class;

    use Moose;
    use MooseX::StrictConstructor;

    has 'size' => ...;

    # then later ...

    # this blows up because color is not a known attribute
    My::Class->new( size => 5, color => 'blue' );

=head1 DESCRIPTION

Simply loading this module makes your constructors "strict". If your
constructor is called with an attribute init argument that your class
does not declare, then it calls C<Moose->throw_error()>. This is a great way
to catch small typos.

=head2 Subverting Strictness

You may find yourself wanting to have your constructor accept a
parameter which does not correspond to an attribute.

In that case, you'll probably also be writing a C<BUILD()> or
C<BUILDARGS()> method to deal with that parameter. In a C<BUILDARGS()>
method, you can simply make sure that this parameter is not included
in the hash reference you return. Otherwise, in a C<BUILD()> method,
you can delete it from the hash reference of parameters.

  sub BUILD {
      my $self   = shift;
      my $params = shift;

      if ( delete $params->{do_something} ) {
          ...
      }
  }

=head1 BUGS

Please report any bugs or feature requests to
C<bug-moosex-strictconstructor@rt.cpan.org>, or through the web
interface at L<http://rt.cpan.org>.  I will be notified, and then
you'll automatically be notified of progress on your bug as I make
changes.

=head1 AUTHOR

Dave Rolsky <autarch@urth.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Dave Rolsky.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


__END__



lib/MooseX/StrictConstructor/Role/Object.pm
package MooseX::StrictConstructor::Role::Object;
BEGIN {
  $MooseX::StrictConstructor::Role::Object::VERSION = '0.15';
}

use Moose::Role;

use namespace::autoclean;

after 'BUILDALL' => sub {
    my $self   = shift;
    my $params = shift;

    my %attrs = (
        __INSTANCE__ => 1,
        map { $_ => 1 }
        grep {defined}
        map  { $_->init_arg() } $self->meta()->get_all_attributes()
    );

    my @bad = sort grep { !$attrs{$_} } keys %{$params};

    if (@bad) {
        Moose->throw_error(
            "Found unknown attribute(s) init_arg passed to the constructor: @bad");
    }

    return;
};

1;

# ABSTRACT: A role which implements a strict constructor for Moose::Object



=pod

=head1 NAME

MooseX::StrictConstructor::Role::Object - A role which implements a strict constructor for Moose::Object

=head1 VERSION

version 0.15

=head1 SYNOPSIS

  Moose::Util::MetaRole::apply_base_class_roles(
      for_class => $caller,
      roles =>
          ['MooseX::StrictConstructor::Role::Object'],
  );

=head1 DESCRIPTION

When you use C<MooseX::StrictConstructor>, your objects will have this
role applied to them. It provides a method modifier for C<BUILDALL()>
from C<Moose::Object> that implements strict argument checking for
your class.

=head1 AUTHOR

Dave Rolsky <autarch@urth.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2011 by Dave Rolsky.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


__END__



lib/MooseX/StrictConstructor/Trait/Class.pm
package MooseX::StrictConstructor::Trait::Class;
{
  $MooseX::StrictConstructor::Trait::Class::VERSION = '0.18';
}

use Moose::Role;

use namespace::autoclean;

use B ();

around new_object => sub {
    my $orig     = shift;
    my $self     = shift;
    my $params   = @_ == 1 ? $_[0] : {@_};
    my $instance = $self->$orig(@_);

    my %attrs = (
        __INSTANCE__ => 1,
        (
            map { $_ => 1 }
            grep {defined}
            map  { $_->init_arg() } $self->get_all_attributes()
        )
    );

    my @bad = sort grep { !$attrs{$_} } keys %$params;

    if (@bad) {
        $self->throw_error(
            "Found unknown attribute(s) init_arg passed to the constructor: @bad"
        );
    }

    return $instance;
};

around _inline_BUILDALL => sub {
    my $orig = shift;
    my $self = shift;

    my @source = $self->$orig();

    my @attrs = (
        '__INSTANCE__ => 1,',
        map { B::perlstring($_) . ' => 1,' }
        grep {defined}
        map  { $_->init_arg() } $self->get_all_attributes()
    );

    return (
        @source,
        'my @bad = sort grep { !$allowed_attrs{$_} } keys %{ $params };',
        'if (@bad) {',
            'Moose->throw_error("Found unknown attribute(s) passed to the constructor: @bad");',
        '}',
    );
} if $Moose::VERSION >= 1.9900;

around _eval_environment => sub {
    my $orig = shift;
    my $self = shift;

    my $env = $self->$orig();

    my %attrs = map { $_ => 1 }
        grep { defined }
        map  { $_->init_arg() }
        $self->get_all_attributes();

    $attrs{__INSTANCE__} = 1;

    $env->{'%allowed_attrs'} = \%attrs;

    return $env;
} if $Moose::VERSION >= 1.9900;

1;

# ABSTRACT: A role to make immutable constructors strict



=pod

=head1 NAME

MooseX::StrictConstructor::Trait::Class - A role to make immutable constructors strict

=head1 VERSION

version 0.18

=head1 DESCRIPTION

This role simply wraps C<_inline_BUILDALL()> (from
C<Moose::Meta::Class>) so that immutable classes have a
strict constructor.

=head1 AUTHOR

Dave Rolsky <autarch@urth.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Dave Rolsky.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


__END__



lib/MooseX/StrictConstructor/Trait/Method/Constructor.pm
package MooseX::StrictConstructor::Trait::Method::Constructor;
{
  $MooseX::StrictConstructor::Trait::Method::Constructor::VERSION = '0.18';
}

use Moose::Role;

use namespace::autoclean;

use B ();

around _generate_BUILDALL => sub {
    my $orig = shift;
    my $self = shift;

    my $source = $self->$orig();
    $source .= ";\n" if $source;

    my @attrs = '__INSTANCE__ => 1,';
    push @attrs, map { B::perlstring($_) . ' => 1,' }
        grep { defined }
        map  { $_->init_arg() } @{ $self->_attributes() };

    $source .= <<"EOF";
my \%attrs = (@attrs);

my \@bad = sort grep { ! \$attrs{\$_} }  keys \%{ \$params };

if (\@bad) {
    Moose->throw_error("Found unknown attribute(s) passed to the constructor: \@bad");
}
EOF

    return $source;
} if $Moose::VERSION < 1.9900;

1;

# ABSTRACT: A role to make immutable constructors strict



=pod

=head1 NAME

MooseX::StrictConstructor::Trait::Method::Constructor - A role to make immutable constructors strict

=head1 VERSION

version 0.18

=head1 DESCRIPTION

This role simply wraps C<_generate_BUILDALL()> (from
C<Moose::Meta::Method::Constructor>) so that immutable classes have a
strict constructor.

=head1 AUTHOR

Dave Rolsky <autarch@urth.org>

=head1 COPYRIGHT AND LICENSE

This software is Copyright (c) 2012 by Dave Rolsky.

This is free software, licensed under:

  The Artistic License 2.0 (GPL Compatible)

=cut


__END__




Makefile.PL

use strict;
use warnings;



use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Make your object constructors blow up on unknown attributes",
  "AUTHOR" => "Dave Rolsky <autarch\@urth.org>",
  "BUILD_REQUIRES" => {
    "Test::Fatal" => 0,
    "Test::Moose" => 0,
    "Test::More" => "0.88"
  },
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "MooseX-StrictConstructor",
  "EXE_FILES" => [],
  "LICENSE" => "artistic_2",
  "NAME" => "MooseX::StrictConstructor",
  "PREREQ_PM" => {
    "B" => 0,
    "Moose" => "0.94",
    "Moose::Exporter" => 0,
    "Moose::Role" => 0,
    "Moose::Util::MetaRole" => 0,
    "namespace::autoclean" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.18",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.56) } ) {
  my $br = delete $WriteMakefileArgs{BUILD_REQUIRES};
  my $pp = $WriteMakefileArgs{PREREQ_PM};
  for my $mod ( keys %$br ) {
    if ( exists $pp->{$mod} ) {
      $pp->{$mod} = $br->{$mod} if $br->{$mod} > $pp->{$mod};
    }
    else {
      $pp->{$mod} = $br->{$mod};
    }
  }
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



